models/glif_psc.h

```diff
@@ -0,0 +1,354 @@
+weight 1.0 results in a peak current of 1 pA at t = tau_syn. On the postsynapic
+side, there can be arbitrarily many synaptic time constants. This can be reached
+by specifying separate receptor ports, each for a different time constant.
+The port number has to match the respective "receptor_type" in the connectors.
+
+The five GLIF models are
+
+GLIF model 1 - Traditional leaky integrate and fire (LIF) model,
+GLIF model 2 - Leaky integrate and fire with biologically defined reset rules
+               model,
+GLIF model 3 - Leaky integrate and fire with after-spike currents model,
+GLIF model 4 - Leaky integrate and fire with biologically defined reset rules
+               and after-spike currents model,
+GLIF model 5 - Leaky integrate and fire with biologically defined reset rules,
+               after-spike currents and a voltage dependent threshold model.
+
+Parameters:
+
+The following parameters can be set in the status dictionary.
+
+V_m        double - Membrane potential in mV
+V_th       double - Instantaneous threshold in mV.
+g          double - Membrane conductance in nS.
+E_L        double - Resting membrane potential in mV.
+C_m        double - Capacitance of the membrane in pF.
+t_ref      double - Duration of refractory time in ms.
+V_reset    double - Reset potential of the membrane in mV (glif 1-3).
+a_spike    double - Threshold addition following spike in mV.
+b_spike    double - Spike-induced threshold time constant in 1/ms.
+a_reset    double - Voltage fraction coefficient following spike.
+b_reset    double - Voltage addition following spike in mV.
+asc_init   double vector - Initial values of after-spike currents in pA.
+k          double vector - After-spike current time constants in 1/ms
+                           (kj in Equation (3) in [1]).
+asc_amps   double vector - After-spike current amplitudes in pA
+                           (deltaIj in Equation (7) in [1]).
+r          double vector - Current fraction following spike coefficients
+                           (fj in Equation (7) in [1]).
+a_voltage  double - Adaptation index of threshold - A 'leak-conductance'
+                    for the voltage-dependent component of the threshold
+                    in 1/ms (av in Equation (4) in [1]).
+b_voltage  double - Voltage-induced threshold time constant - Inverse of which
+                    is the time constant of the voltage-dependent component
+                    of the threshold in 1/ms (bv in Equation (4) in [1]).
+tau_syn    double vector - Rise time constants of the synaptic alpha function
+                           in ms.
+glif_model string - glif model type:
+                    "lif"/"glif_lif"/"1" - GLIF model 1,
+                    "lif_r"/"glif_lif_r"/"2" - GLIF model 2,
+                    "lif_asc"/"glif_lif_asc"/"3" - GLIF model 3,
+                    "lif_r_asc"/"glif_lif_r_asc"/"4" - GLIF model 4, or
+                    "lif_r_asc_a"/"glif_lif_r_asc_a"/"5" - GLIF model 5.
```

Comment:
I would suggest to allow only the "lif_..." names. The "glif_" is redundant and numbers are uninformative and will make simulation scripts difficult to read in the long term. I am also wondering if we could drop "lif" from the names, since it is redundant.

Maybe we should consider a different scheme: How about having a parameter "mechanism" (I am open for better suggestions) which is an array of strings, expressing which mechanisms to incorporate:

- 1: `/mechanisms []`
- 2: `/mechanisms [ /spike_dependent_threshold ]`
- 3: `/mechanisms [ /after_spike_currents  ]`
- 4: `/mechanisms [ /spike_dependent_threshold /after_spike_currents  ]`
- 5: `/mechanisms [ /spike_dependent_threshold /after_spike_currents /adapting_threshold ]`
 
This is more verbose, but I think it will make reading any script using the models much more comprehensible a few years down the line and will also avoid user errors by confusing, e.g., model 3 and 4. Internally in the model this could be mapped to model numbers 1-5 and unsupported combinations could be forbidden.

For an example of how to read out such a list of labels, see the code handling `record_to` in [recording_device.cpp](https://github.com/nest/nest-simulator/blob/8adec3cd2bb8025ca07c91f8e281ecb71f13658f/nestkernel/recording_device.cpp#L241); the `/` are in SLI only, in Python this would be strings.

Reply:
I have created three bool parameters: spike_dependent_threshold, after_spike_currents and adapting_threshold, for model mechanisms setting. This should be much straightforward and more verbose for users. Checking of combination of these three parameters has also been added for only supporting the five glif models. Other combinations of input will throw BadProperty. Updates will be pushed once done with other comments.